src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -0,0 +1,100 @@
+    [Shared]
+    [ExportLspMethod(LSP.Methods.TextDocumentCompletionName)]
+    internal class CompletionHandler : IRequestHandler<LSP.CompletionParams, object>
+    {
+        public async Task<object> HandleRequestAsync(Solution solution, LSP.CompletionParams request, LSP.ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
+        {
+            var document = solution.GetDocumentFromURI(request.TextDocument.Uri);
+            if (document == null)
+            {
+                return Array.Empty<LSP.CompletionItem>();
+            }
+
+            var position = await document.GetPositionFromLinePositionAsync(ProtocolConversions.PositionToLinePosition(request.Position), cancellationToken).ConfigureAwait(false);
+
+            var completionService = document.Project.LanguageServices.GetService<CompletionService>();
+            var list = await completionService.GetCompletionsAsync(document, position, cancellationToken: cancellationToken).ConfigureAwait(false);
+            if (list == null)
+            {
+                return Array.Empty<LSP.CompletionItem>();
+            }
+
+            var lspClientCapability = clientCapabilities.HasVisualStudioLspCapability();
+
+            return list.Items.Select(item =>
+            {
+                LSP.CompletionItem completionItem;
+                if (lspClientCapability == true)
+                {
+                    completionItem = CreateCompletionItem<LSP.VSCompletionItem>(request, item);
+                    ((LSP.VSCompletionItem)completionItem).Icon = new ImageElement(item.Tags.GetFirstGlyph().GetImageId());
+                }
+                else
+                {
+                    completionItem = CreateCompletionItem<RoslynCompletionItem>(request, item);
+                    ((RoslynCompletionItem)completionItem).Tags = item.Tags.ToArray();
+                }
+                return completionItem;
+            }).ToArray();
+
+            // local functions
+            static TCompletionItem CreateCompletionItem<TCompletionItem>(LSP.CompletionParams request, CompletionItem item) where TCompletionItem : LSP.CompletionItem, new()
+            {
+                return new TCompletionItem
+                {
+                    Label = item.DisplayText,
+                    InsertText = item.Properties.ContainsKey("InsertionText") ? item.Properties["InsertionText"] : item.DisplayText,
+                    SortText = item.SortText,
+                    FilterText = item.FilterText,
+                    Kind = GetCompletionKind(item.Tags),
+                    Data = new CompletionResolveData { CompletionParams = request, DisplayText = item.DisplayText }
+                };
+            }
+        }
+
+        private static LSP.CompletionItemKind GetCompletionKind(ImmutableArray<string> tags)
+        {
+            foreach (var tag in tags)
+            {
+                if (Enum.TryParse<LSP.CompletionItemKind>(tag, out var kind))
```

Comment:
how about we use cache to remember tag <-> value once converted? I think tag is known set.

Reply:
or we could create static map for CompletionItemKind to remove enum parsing.